### PR TITLE
Events on state changed

### PR DESCRIPTION
**Pull request by Ferdudas97**
Audit global snapshot when it is created

### CODE REVIEW

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotUpdater.kt [hunk 1]
```diff
@@ -117,14 +122,20 @@ class SnapshotUpdater(
...
                     xdsSnapshot = lastXdsSnapshot
                 )
                 globalSnapshot = updateResult
                 updateResult
+            }.scan { previous: UpdateResult?, actual: UpdateResult? ->
+                val emptyUpdateResult = emptyUpdateResult()
+                snapshotChangeAuditor.audit(previous ?: emptyUpdateResult, actual ?: emptyUpdateResult)
+                    .publishOn(globalSnapshotAuditScheduler)
+                    .subscribeOn(globalSnapshotAuditScheduler)
+                    .subscribe().apply { auditDisposables.add(this) }
```
Comment (malfoj89) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotUpdater.kt:137]: why we need to use disposable here?

Comment (malfoj89) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotUpdater.kt:137]: imho subscribeon and subscribe will be enough u do not need to do close or publishOn here

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/ControlPlane.kt [hunk 2]
```diff
@@ -85,39 +87,29 @@ class ControlPlane private constructor(
...
+        var grpcServerExecutor: Executor = buildThreadPoolExecutor()
+        // unbounded executor - netty will only use configured number of threads
+        // (by nioEventLoopThreadCount property or default netty value: <number of CPUs> * 2)
+        var nioEventLoopExecutor: Executor = newMeteredCachedThreadPool("grpc-worker-event-loop")
+        var executorGroup: ExecutorGroup = buildExecutorGroup()
+        var globalSnapshotExecutor: Executor = newMeteredFixedThreadPool(
+            "snapshot-update",
+            properties.server.globalSnapshotUpdatePoolSize
+        )
+        var globalSnapshotAuditExecutor: Executor = newMeteredFixedThreadPool(
```
Comment (gawarz) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/ControlPlane.kt:99]: This is valid approach for our usage, but if sb use method withGlobalSnapshotAuditExecutor our builder will create additional ThreadPool. 

I prefer lazy approach

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotChangeAuditor.kt [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+import reactor.core.publisher.Mono
+
+interface SnapshotChangeAuditor {
+    fun audit(previousUpdateResult: UpdateResult, actualUpdateResult: UpdateResult): Mono<Any>
+}
+
+class NoopSnapshotChangeAuditor: SnapshotChangeAuditor {
+    override fun audit(previousUpdateResult: UpdateResult, actualUpdateResult: UpdateResult): Mono<Any> {
+        return Mono.empty()
```
Comment (gawarz) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotChangeAuditor.kt:11]: 👍

**Review by malfoj89 — APPROVED** [review]

**Review by gawarz — APPROVED** [review]

**Review by malfoj89 — APPROVED** [review]

**Pull request merged**